template/{% if use_git and dev_platform == 'GitHub' %}.github{% endif %}/workflows/release.yml

```
@@ -1,53 +1,74 @@
+    # rather than forks
+    # Environment is encouraged for increased security
+    environment: build
     steps:
-    - uses: actions/checkout@v4
-
-    - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v5
-      with:
-        python-version: ${{ matrix.python-version }}
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          # This fetch element is only important if you are use SCM based
+          # versioning (that looks at git tags to gather the version)
+          fetch-depth: 100
+            # Need the tags so that setuptools-scm can form a valid version number
+      - name: Fetch git tags
+        run: git fetch origin 'refs/tags/*:refs/tags/*'
 
-    - name: Install hatch
-      uses: pypa/hatch@install
+      - name: Setup Python
+        uses: actions/setup-python@v5
+        with:
+          # You can modify what version of Python you want to use for your release
+          python-version: "3.11"
 
-    - name: Build package
-      run: hatch build
+      # Security recommends we should pin deps. Should we pin the workflow version?
+      - name: Install hatch
+        uses: pypa/hatch@install
 
-    # We rely on a trusted publisher configuration being present on PyPI,
-    # see https://docs.pypi.org/trusted-publishers/.
-    - name: Publish to PyPI
-      uses: pypa/gh-action-pypi-publish@release/v1
+      - name: Build package using Hatch
+        run: |
+          hatch build
+          echo ""
+          echo "Generated files:"
+          ls -lh dist/
 
-    - name: GH release
-      uses: softprops/action-gh-release@v2
-      with:
-        body: >
-          Please see
-          https://github.com/${{ github.repository }}/blob/${{ github.ref_name }}/CHANGELOG.md
-          for the full release notes.
-        draft: false
-        prerelease: false
+      # Store an artifact of the build to use in the publish step below
+      - name: Store the distribution packages
+        uses: actions/upload-artifact@v4
+        with:
+          name: python-package-distributions
+          path: dist/
+  publish:
+    name: >-
+      Publish Python 🐍 distribution 📦 to PyPI
+    # Modify the repo name below to be your project's repo name.
+    if: github.repository_owner == 'pyopensci'
+    needs: 
+      - build
+    runs-on: ubuntu-latest
+    # Environment required here for trusted publisher
+    environment:
+      name: pypi
+      # Modify the url to be the name of your package
+      url: https://pypi.org/p/yourPackage
+    permissions:
+      id-token: write  # this permission is mandatory for PyPI publishing
+    steps:
+      - name: Download dists
+        uses: actions/download-artifact@v4
+        with:
+          name: python-package-distributions
+          path: dist/
+      - name: Publish package to PyPI
+        # Only publish to real PyPI on release
+        if: github.event_name == 'release'
```

Is this line really needed since the whole workflow only triggers on release?

---

I would suggest including it, as (in my opinion) CD/release-related workflows like this are also supposed to run on a periodic basis using a CRON job (with a `schedule:` trigger) to see if the build part does not have any issues. This applies much more for packages with compiled codebases where a host of errors from the entirety of their toolchains and supported platforms are the norm, but is also applicable for almost any release workflow, i.e., even for a pure Python package – its dependencies can break, build failures can occur, and so on.

If at all, I would suggest making it more stringent, like this:

```suggestion
        if: github.event_name == 'release' && github.event.action == 'published'
```

in the hope of making a release-oriented developer more careful (similar to my point above) to check things twice.

---

so this is a triple check. i'm going to call in @webknjaz -  Sviatoslav this is a template for pyOpenSci that people can use to quickly create a package structure. I Updated the GitHub action part to reflect what PyPA suggests in terms of security. Here we are having a small friendly debate around how many event checks should go into the build. I'm curious what your take on this is?

More can't hurt, I would presume. it just adds a bit of complexity to the build to explain to people. So i want to make sure we keep things simple yet safe! thank you!!